### PR TITLE
[PoC] `accelerate` support for trl

### PR DESCRIPTION
**Pull request by younesbelkada**
Hey there!

Let's see how we can integrate `accelerate` to enable larger models training (up to 10B scale). Here is a draft version on how we could potentially approach that. 

Firstly, I didn't wrote tests yet, but for now, I think that having example scripts on `examples/` and running them while developing seems to be the right approach. In the near future we'll need to add tests. I also think that we should develop the code outside `nbdev`

Secondly, I think that we should be able to support more models instead of writing a new file per architecture. The will be addressed too. 

To sumarize, this PR addresses the following:

## 1- `accelerate` integration for naive training using `accelerate`

For now this only supports the native `accelerate` integration. It includes automatic mixed precision training and GPU device assignment. I need to test multi-GPU setup & Data Parallelism

## 2- Support for `xxxForCausalLM` architecture

Anyone can start experimenting using any of the `xxxForCausalLM` architecture, this would include any of the `Bloom`, `OPT`, etc. 
The API looks like the following for now and can be adapted:
```
from transformers import AutoModelForCausalLM
from trl.models import AutoRegressiveLMWithValueHead

model = AutoModelForCausalLM.from_pretrained("bigscience/bloom-560m")
ref_model = AutoRegressiveLMWithValueHead(model)
```

## 3- Support latest version of `transformers`

this PR adds also slight modifications to support latest `transformers` version.

## 4- Community orientation

I believe this repo should be a center point for anyone on the community that wants to experiment on hfrl. I think that we should educate users and contributors on how to implement a new training method. For now to add a new training method a user should:
- Implement its own `xxxTrainer` class that inherits from `BaseTrainer` class
- add a clear example on `examples/`
That's it!

## 5- How am I testing the implementation?

For now I am running `accelerate launch example/accelerate-ppo.py` and observing the training dynamics of the model and assert that it has more or less the same behavior than the reference one: https://wandb.ai/lvwerra/trl-showcase/runs/1jtvxb1m?workspace= 
Usually after 6-7 steps the reward starts to skyrocket.

## What is next?

- [ ] discuss how we can improve the API
- [ ] remove dependency for `nbdev`
- [x] broader support for more architectures
- [ ] test multi-GPU setup (DP) 
- [ ] Implement tests

cc @lvwerra

### CODE REVIEW

**Comment by younesbelkada** [comment]
For now I am testing my implementation with `accelerate launch example/ppo-accelerate.py`

**Comment by younesbelkada** [comment]
Regarding tests, this is tricky but from what I can see we can for now:
- Test if all trainers respects the inheritance from `BaseTrainer` (by checking if all the needed functions are implemented)
- Test if all models work as expected (thinking of `generate` method) and if we can in fact support all `xxxForCausalLM` architectures as claimed above. From what I can see, as long as the model has a proper `generate` method the PPOTrainer should work

File: trl/trainer/accelerate_ppo.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+        """
+        super().__init__(self.default_params)
+
+        # Step 1: Initialize Accelerator
+        self.accelerator = Accelerator(log_with="wandb")
+        self.config.update(config)
+
+        # Step 2: Initialize model, tokenizer and dataset
+        self._build_models_and_tokenizer()
```
Comment (lvwerra) [inline trl/trainer/accelerate_ppo.py:80]: i think we can follow a similar logic to what we do in the Trainer/Evaluator classes: the models can be either provided as torch.models or as strings in which case we load them.

File: trl/trainer/accelerate_ppo.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+                                       non_score_reward=non_score_reward, train_stats=train_stats,
+                                       kl_coef=self.kl_ctl.value)
+        stats = stats_to_np(stats)
+        timing['time/ppo/calc_stats'] = time.time()-t
+
+        self.kl_ctl.update(stats['objective/kl'], self.config['batch_size'])
+
+        timing['time/ppo/total'] = time.time()-t0
+        stats.update(timing)
+        return stats
```
Comment (lvwerra) [inline trl/trainer/accelerate_ppo.py:232]: since you initialize the W&B logger with accelerate we can probably just pass all the stats to it internally and don't put the burden on the user.

Also, i think we can make the type of logger a kwarg that's passed to PPO so we don't require users to use W&B and they could also setup other loggers.

**Review by lvwerra — COMMENTED** [review]
Awesome work @younesbelkada! I left a few comments but I think figuring out all the details for the proposed changes will take some time so what do you think about breaking it down a bit:

- remove `nbdev` and setup a CI
- PR for the the `AutoModelForCasualLMWithValueHead` (we can easily test if this works as expected)
- I think we only want the Accelerate trainer (which should be equivalent to what's there now if you just run accelerate on a single machine).
- The two places where we would benefit the most from data parallelism is `generation` and `step` - let's think about how we can achieve in the most elegant way.

I also need to read up a bit on current strategies. Maybe there are otherways how to optimize the training loops (e.g. do ref and trained model to be different or could it be the same model with two heads).

**Comment by younesbelkada** [comment]
Closing in favor of https://github.com/lvwerra/trl/pull/58
